File: test/integration/PrimitiveEquivalenceDriver.go
//===----------------------------------------------------------------------===//
///
/// @file
/// Isolated cross-language runtime-primitive equivalence driver (Go).
///
/// Reads the shared vector file (os.Args[1]) and exercises each vector directly
/// against the Go runtime primitives, checking every primitive direction on its
/// own so paired bugs cannot mask each other. Prints "PROCESSED <n>" on success
/// and exits non-zero on the first mismatch. The C and Rust drivers consume the
/// exact same file.
///
//===----------------------------------------------------------------------===//

package main

import (
	"bufio"
	"bytes"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"

	rt "github.com/thirtytwobits/llvm-dsdl/runtime/go"
)

func parseHexBytes(s string) ([]byte, error) {
	if len(s)%2 != 0 {
		return nil, fmt.Errorf("odd-length hex: %q", s)
	}
	out := make([]byte, len(s)/2)
	for i := 0; i < len(s); i += 2 {
		v, err := strconv.ParseUint(s[i:i+2], 16, 8)
		if err != nil {
			return nil, err
		}
		out[i/2] = byte(v)
	}
	return out, nil
}

func hexU64(s string) uint64 {
	v, _ := strconv.ParseUint(s, 16, 64)
	return v
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage: driver <vectors-file>")
		os.Exit(2)
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "cannot open vectors file: %s\n", os.Args[1])
		os.Exit(2)
	}
	defer f.Close()

	processed := 0
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()
		if i := strings.IndexByte(line, '#'); i >= 0 {
			line = line[:i]
		}
		fields := strings.Fields(line)
		if len(fields) == 0 {
			continue
		}
		op := fields[0]
		fail := func(detail string) {
			fmt.Fprintf(os.Stderr, "MISMATCH %s: %s\n", op, detail)
			os.Exit(1)
		}

		switch op {
		case "f16pack":
			if len(fields) != 3 {
				fail("arity")
			}
			got := rt.Float16Pack(math.Float32frombits(uint32(hexU64(fields[1]))))
			want := uint16(hexU64(fields[2]))
			if got != want {
				fail(fmt.Sprintf("in=%s want=%04x got=%04x", fields[1], want, got))
			}
		case "f16unpack":
			if len(fields) != 3 {
				fail("arity")
			}
			got := math.Float32bits(rt.Float16Unpack(uint16(hexU64(fields[1]))))
			want := uint32(hexU64(fields[2]))
			if got != want {
				fail(fmt.Sprintf("in=%s want=%08x got=%08x", fields[1], want, got))
			}
		case "geti", "getu":
			if len(fields) != 6 {
				fail("arity")
			}
			typeBits, _ := strconv.Atoi(fields[1])
			lenBits, _ := strconv.Atoi(fields[2])
			offBits, _ := strconv.Atoi(fields[3])
			buf, e := parseHexBytes(fields[4])
			if e != nil {
				fail("bad buffer hex")
			}
			want := hexU64(fields[5])
			var got uint64
			if op == "geti" {
				switch typeBits {
				case 8:
					got = uint64(int64(rt.GetI8(buf, offBits, uint8(lenBits))))
				case 16:
					got = uint64(int64(rt.GetI16(buf, offBits, uint8(lenBits))))
				case 32:
					got = uint64(int64(rt.GetI32(buf, offBits, uint8(lenBits))))
				case 64:
					got = uint64(rt.GetI64(buf, offBits, uint8(lenBits)))
				default:
					fail("bad type_bits")
				}
			} else {
				switch typeBits {
				case 8:
					got = uint64(rt.GetU8(buf, offBits, uint8(lenBits)))
				case 16:
					got = uint64(rt.GetU16(buf, offBits, uint8(lenBits)))
				case 32:
					got = uint64(rt.GetU32(buf, offBits, uint8(lenBits)))
				case 64:
					got = rt.GetU64(buf, offBits, uint8(lenBits))
				default:
					fail("bad type_bits")
				}
			}
			if got != want {
				fail(fmt.Sprintf("type=%d len=%d off=%d buf=%s want=%016x got=%016x",
					typeBits, lenBits, offBits, fields[4], want, got))
			}
		case "copybits":
			if len(fields) != 7 {
				fail("arity")
			}
			dstOff, _ := strconv.Atoi(fields[1])
			lenBits, _ := strconv.Atoi(fields[2])
			srcOff, _ := strconv.Atoi(fields[3])
			src, e1 := parseHexBytes(fields[4])
			dst, e2 := parseHexBytes(fields[5])
			want, e3 := parseHexBytes(fields[6])
			if e1 != nil || e2 != nil || e3 != nil || len(dst) != len(want) {
				fail("bad buffer hex / length")
			}
			rt.CopyBits(dst, dstOff, lenBits, src, srcOff)
			if !bytes.Equal(dst, want) {
				fail(fmt.Sprintf("dstoff=%d len=%d srcoff=%d got=%x", dstOff, lenBits, srcOff, dst))
			}
		default:
			fmt.Fprintf(os.Stderr, "UNRECOGNIZED vector: %s\n", line)
			os.Exit(1)
		}
		processed++
	}
	if err := sc.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "read error: %v\n", err)
		os.Exit(2)
	}

	fmt.Printf("PROCESSED %d\n", processed)
	fmt.Println("Go primitive equivalence PASS")
}

File: test/integration/PrimitiveEquivalenceDriver.rs
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//
//
// Isolated cross-language runtime-primitive equivalence driver (Rust).
//
// Reads the shared vector file (argv[1]) and exercises each vector directly
// against the Rust runtime primitives, checking every primitive direction on
// its own so paired bugs cannot mask each other. Prints "PROCESSED <n>" on
// success and exits non-zero on the first mismatch. The C and Go drivers consume
// the exact same file.
//
// The build supplies a sibling `dsdl_runtime` module that points at the real
// `runtime/rust/dsdl_runtime.rs`.
//
//===----------------------------------------------------------------------===//

#![allow(clippy::unusual_byte_groupings)]

mod dsdl_runtime;

use std::io::BufRead;
use std::process::exit;

fn parse_hex_bytes(s: &str) -> Result<Vec<u8>, String> {
    if s.len() % 2 != 0 {
        return Err(format!("odd-length hex: {s}"));
    }
    let mut out = Vec::with_capacity(s.len() / 2);
    let bytes = s.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        let hi = &s[i..i + 2];
        out.push(u8::from_str_radix(hi, 16).map_err(|e| e.to_string())?);
        i += 2;
    }
    Ok(out)
}

fn hex_u64(s: &str) -> u64 {
    u64::from_str_radix(s, 16).unwrap_or(0)
}

fn fail(op: &str, detail: &str) -> ! {
    eprintln!("MISMATCH {op}: {detail}");
    exit(1);
}

fn main() {
    let path = match std::env::args().nth(1) {
        Some(p) => p,
        None => {
            eprintln!("usage: driver <vectors-file>");
            exit(2);
        }
    };
    let file = match std::fs::File::open(&path) {
        Ok(f) => f,
        Err(_) => {
            eprintln!("cannot open vectors file: {path}");
            exit(2);
        }
    };

    let mut processed: usize = 0;
    for line in std::io::BufReader::new(file).lines() {
        let line = line.unwrap_or_default();
        let line = match line.find('#') {
            Some(idx) => &line[..idx],
            None => &line[..],
        };
        let f: Vec<&str> = line.split_whitespace().collect();
        if f.is_empty() {
            continue;
        }
        let op = f[0];

        match op {
            "f16pack" => {
                if f.len() != 3 {
                    fail(op, "arity");
                }
                let got = dsdl_runtime::float16_pack(f32::from_bits(hex_u64(f[1]) as u32));
                let want = hex_u64(f[2]) as u16;
                if got != want {
                    fail(op, &format!("in={} want={want:04x} got={got:04x}", f[1]));
                }
            }
            "f16unpack" => {
                if f.len() != 3 {
                    fail(op, "arity");
                }
                let got = dsdl_runtime::float16_unpack(hex_u64(f[1]) as u16).to_bits();
                let want = hex_u64(f[2]) as u32;
                if got != want {
                    fail(op, &format!("in={} want={want:08x} got={got:08x}", f[1]));
                }
            }
            "geti" | "getu" => {
                if f.len() != 6 {
                    fail(op, "arity");
                }
                let type_bits: u32 = f[1].parse().unwrap_or(0);
                let len_bits: u8 = f[2].parse().unwrap_or(0);
                let off_bits: usize = f[3].parse().unwrap_or(0);
                let buf = parse_hex_bytes(f[4]).unwrap_or_else(|_| fail(op, "bad buffer hex"));
                let want = hex_u64(f[5]);
                let got: u64 = if op == "geti" {
                    match type_bits {
                        8 => dsdl_runtime::get_i8(&buf, off_bits, len_bits) as i64 as u64,
                        16 => dsdl_runtime::get_i16(&buf, off_bits, len_bits) as i64 as u64,
                        32 => dsdl_runtime::get_i32(&buf, off_bits, len_bits) as i64 as u64,
                        64 => dsdl_runtime::get_i64(&buf, off_bits, len_bits) as u64,
                        _ => fail(op, "bad type_bits"),
                    }
                } else {
                    match type_bits {
                        8 => dsdl_runtime::get_u8(&buf, off_bits, len_bits) as u64,
                        16 => dsdl_runtime::get_u16(&buf, off_bits, len_bits) as u64,
                        32 => dsdl_runtime::get_u32(&buf, off_bits, len_bits) as u64,
                        64 => dsdl_runtime::get_u64(&buf, off_bits, len_bits),
                        _ => fail(op, "bad type_bits"),
                    }
                };
                if got != want {
                    fail(
                        op,
                        &format!(
                            "type={type_bits} len={len_bits} off={off_bits} buf={} want={want:016x} got={got:016x}",
                            f[4]
                        ),
                    );
                }
            }
            "copybits" => {
                if f.len() != 7 {
                    fail(op, "arity");
                }
                let dst_off: usize = f[1].parse().unwrap_or(0);
                let len_bits: usize = f[2].parse().unwrap_or(0);
                let src_off: usize = f[3].parse().unwrap_or(0);
                let src = parse_hex_bytes(f[4]).unwrap_or_else(|_| fail(op, "bad buffer hex"));
                let mut dst = parse_hex_bytes(f[5]).unwrap_or_else(|_| fail(op, "bad buffer hex"));
                let want = parse_hex_bytes(f[6]).unwrap_or_else(|_| fail(op, "bad buffer hex"));
                if dst.len() != want.len() {
                    fail(op, "dst/want length mismatch");
                }
                dsdl_runtime::copy_bits(&mut dst, dst_off, len_bits, &src, src_off);
                if dst != want {
                    fail(
                        op,
                        &format!("dstoff={dst_off} len={len_bits} srcoff={src_off} got={dst:02x?}"),
                    );
                }
            }
            _ => {
                eprintln!("UNRECOGNIZED vector: {line}");
                exit(1);
            }
        }
        processed += 1;
    }

    println!("PROCESSED {processed}");
    println!("Rust primitive equivalence PASS");
}
